File: ios/Sources/WhatsAppGateway.swift
import Foundation
import Whatsbridge

private final class WhatsEventListener: NSObject, WhatsbridgeListener {
    var handler: ((String) -> Void)?
    func onEvent(_ eventJSON: String?) { if let eventJSON { handler?(eventJSON) } }
}

final class WhatsAppGateway: MessagingGateway {
    var event: ((GatewayEvent) -> Void)?
    private let accountID: UUID
    private let root: URL
    private let listener = WhatsEventListener()
    private var manager: WhatsbridgeManager?

    init(accountID: UUID, root: URL) {
        self.accountID = accountID
        self.root = root
        listener.handler = { [weak self] in self?.receive($0) }
    }

    func connect(account: Account, phone: String) async throws {
        if manager == nil {
            try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true,
                                                    attributes: [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication])
            manager = try WhatsbridgeNewManager(root.path, listener)
        }
        try manager?.connect()
    }

    func loadChats() async throws {
        guard let raw = try manager?.listChats(), let data = raw.data(using: .utf8),
              let rows = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return }
        for row in rows { emitChat(row) }
    }

    func loadHistory(chatID: String, limit: Int) async throws { try manager?.loadHistory(chatID, limit: limit) }
    func sendText(chatID: String, text: String) async throws -> String {
        guard let manager else { throw StoreError.configuration("WhatsApp не подключён") }
        return try manager.sendText(chatID, text: text)
    }
    func logOut() async throws { try manager?.logout() }
    func close() { manager?.close(); manager = nil }

    private func receive(_ json: String) {
        guard let data = json.data(using: .utf8),
              let row = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = row["type"] as? String else { return }
        switch type {
        case "auth":
            let raw = row["state"] as? String ?? "error"
            let state: AuthState = raw == "ready" ? .ready : (raw == "connecting" ? .connecting : .loggedOut)
            event?(.auth(accountID: accountID, state: state, detail: row["detail"] as? String ?? ""))
        case "qr": event?(.qr(accountID: accountID, value: row["qr"] as? String ?? ""))
        case "chat": emitChat(row)
        case "text":
            event?(.text(accountID: accountID, chatID: row["chat_id"] as? String ?? "",
                         chatTitle: row["chat_title"] as? String ?? "",
                         providerID: row["message_id"] as? String ?? "", text: row["text"] as? String ?? "",
                         timestamp: Date(timeIntervalSince1970: (row["timestamp"] as? NSNumber)?.doubleValue ?? 0),
                         outgoing: row["outgoing"] as? Bool ?? false))
        case "error": event?(.error(accountID: accountID, message: row["detail"] as? String ?? "WhatsApp error"))
        default: break
        }
    }

    private func emitChat(_ row: [String: Any]) {
        let id = row["chat_id"] as? String ?? row["id"] as? String ?? ""
        guard !id.isEmpty else { return }
        event?(.chat(Chat(accountID: accountID, remoteID: id,
                          title: row["chat_title"] as? String ?? row["title"] as? String ?? id,
                          lastText: row["last_text"] as? String ?? "",
                          lastAt: Date(timeIntervalSince1970: (row["timestamp"] as? NSNumber)?.doubleValue
                                   ?? (row["last_at"] as? NSNumber)?.doubleValue ?? 0),
                          unread: (row["unread"] as? NSNumber)?.intValue ?? 0, secureState: .none)))
    }
}

File: android/app/src/main/java/dev/msgcrypt/app/provider/WhatsAppGateway.java
package dev.msgcrypt.app.provider;

import dev.msgcrypt.app.model.Account;
import dev.msgcrypt.app.model.AuthState;
import dev.msgcrypt.app.model.Chat;
import dev.msgcrypt.app.model.SecureState;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class WhatsAppGateway extends UnavailableGateway {
    private final File root;
    private final ExecutorService io = Executors.newSingleThreadExecutor();
    private volatile Object manager;
    private volatile String accountId;

    public WhatsAppGateway(File root) {
        this.root = root;
    }

    @Override
    public CompletableFuture<Void> connect(Account account, String ignoredPhone) {
        accountId = account.id;
        return run(() -> {
            ensureManager();
            call(manager, "connect");
            return null;
        });
    }

    @Override
    public CompletableFuture<Void> loadChats(String accountId) {
        return run(() -> {
            ensureManager();
            String json = String.valueOf(call(manager, "listChats"));
            JSONArray list = new JSONArray(json);
            for (int i = 0; i < list.length(); i++) {
                JSONObject item = list.getJSONObject(i);
                emitChat(item.optString("id"), item.optString("title"), item.optString("last_text"),
                        item.optLong("last_at"), item.optInt("unread"));
            }
            return null;
        });
    }

    @Override
    public CompletableFuture<Void> loadHistory(String accountId, String chatId, int limit) {
        return run(() -> { ensureManager(); call(manager, "loadHistory", chatId, (long) limit); return null; });
    }

    @Override
    public CompletableFuture<String> sendText(String accountId, String chatId, String text) {
        return run(() -> { ensureManager(); return String.valueOf(call(manager, "sendText", chatId, text)); });
    }

    @Override
    public CompletableFuture<Void> logOut(String accountId) {
        return run(() -> { if (manager != null) call(manager, "logout"); return null; });
    }

    @Override
    public void close() {
        Object value = manager;
        manager = null;
        if (value != null) {
            try { call(value, "close"); } catch (Exception ignored) {}
        }
        io.shutdownNow();
    }

    private void ensureManager() throws Exception {
        if (manager != null) return;
        synchronized (this) {
            if (manager != null) return;
            Class<?> listenerClass = Class.forName("dev.msgcrypt.bridge.whatsbridge.Listener");
            Object proxy = Proxy.newProxyInstance(listenerClass.getClassLoader(), new Class<?>[]{listenerClass},
                    (object, method, args) -> {
                        if (method.getName().equalsIgnoreCase("onEvent") && args != null && args.length == 1) {
                            onBridgeEvent(String.valueOf(args[0]));
                        }
                        return null;
                    });
            Class<?> entry = Class.forName("dev.msgcrypt.bridge.whatsbridge.Whatsbridge");
            manager = call(entry, "newManager", root.getAbsolutePath(), proxy);
        }
    }

    private void onBridgeEvent(String raw) {
        try {
            JSONObject value = new JSONObject(raw);
            String type = value.optString("type");
            GatewayListener target = listener;
            if (target == null) return;
            switch (type) {
                case "auth":
                    target.onAuthState(accountId, auth(value.optString("state")), value.optString("detail"));
                    break;
                case "qr":
                    target.onQrCode(accountId, value.getString("qr"));
                    break;
                case "chat":
                    emitChat(value.optString("chat_id"), value.optString("chat_title"),
                            value.optString("last_text"), value.optLong("timestamp"), value.optInt("unread"));
                    break;
                case "text":
                    target.onText(accountId, value.getString("chat_id"), value.optString("chat_title"),
                            value.optString("message_id"), value.getString("text"), value.optLong("timestamp"),
                            value.optBoolean("outgoing"));
                    break;
                case "error":
                    target.onError(accountId, value.optString("detail", "WhatsApp error"), null);
                    break;
            }
        } catch (Exception error) {
            GatewayListener target = listener;
            if (target != null) target.onError(accountId, "Некорректное событие WhatsMeow", error);
        }
    }

    private void emitChat(String id, String title, String lastText, long lastAt, int unread) {
        GatewayListener target = listener;
        if (target != null && id != null && !id.isBlank()) {
            target.onChat(new Chat(accountId, id, title, lastText, lastAt, unread, SecureState.NONE));
        }
    }

    private static AuthState auth(String value) {
        if ("ready".equals(value)) return AuthState.READY;
        if ("logged_out".equals(value)) return AuthState.LOGGED_OUT;
        if ("qr".equals(value)) return AuthState.WAITING_QR;
        return AuthState.CONNECTING;
    }

    private <T> CompletableFuture<T> run(ThrowingSupplier<T> action) {
        return CompletableFuture.supplyAsync(() -> {
            try { return action.get(); }
            catch (Exception error) { throw new java.util.concurrent.CompletionException(error); }
        }, io);
    }

    private static Object call(Object target, String name, Object... args) throws Exception {
        Class<?> type = target instanceof Class<?> ? (Class<?>) target : target.getClass();
        for (Method method : type.getMethods()) {
            if (method.getName().equalsIgnoreCase(name) && method.getParameterCount() == args.length) {
                return method.invoke(target instanceof Class<?> ? null : target, args);
            }
        }
        throw new NoSuchMethodException(type.getName() + "." + name + "/" + args.length);
    }

    private interface ThrowingSupplier<T> { T get() throws Exception; }
}

